### DISCUSSION TITLE
Add helper that automatically gets services according to type hints?

### DISCUSSION THREAD

**Discussion by hynek** [hynek/svcs#94] [Ideas]
As it just came up in #86, I use a pattern similar to this all the time:

```python
class Inner:
    ...

@dataclass
class Wrapper:
    foo: Inner

registry.register_factory(Inner, Inner)
registry.register_factory(Wrapper, lambda svcs_container: Wrapper(svcs_container.get(Inner)))
```

*svcs* will always be non-magical by default, but I guess nothing speaks against adding a helper that makes the lambda superfluous by looking at the type hints of the factory and calling `svcs_container.(a)get` on the required types automatically:

```python
registry.register_factory(Inner, Inner)
registry.register_factory(Wrapper, svcs.inspect_and_get(Wrapper))
```

Since this would only be a helper, it wouldn't also have the pressure to implement all edge cases because the escape hatch of just writing your own lambda is always there.

**Comment by pauleveritt** [comment]
FWIW I have a registry project with the kind of injection you describe. It’s critical for a larger effort I have in mind. 

I’m investigating re-homing my stuff atop svcs… going through a gap analysis now to see which of my darlings (features) to kill.

**Comment by proofit404** [comment]
To my understanding we have a couple of corner cases here:

1. Please, please, please, do not call it `inspect_and_get`
2. We need to have sync and async versions of same function
3. handle keyword arguments defaults
    ```python
    class Srvc:
        a: DepA
        b: DepB = DepB()
    ```
    This would be correct case if registry defines only `DepA`.
4. Passed class could be without type annotations

One benefit I see for factory builder over registry method is ability to customize factory:

```python
fctr = make_factory(Srvc)

def factory(svcs_container):
    with fctr(svcs_container) as instance:
        yield instance

registry.register(Srvc, factory)
```
In case our service has it's own life circle an supposed to be used as context manager.

**Reply by hynek** [reply]
> To my understanding we have a couple of corner cases here:
> 
> 1. Please, please, please, do not call it `inspect_and_get`
> 2. We need to have sync and async versions of same function

Could both be avoided by making it a class… say… `AutoSvc(Wrapper)` and adding special logic into _svcs_'s core if necessary?

> 4. Passed class could be without type annotations
> 
> One benefit I see for factory builder over registry method is ability to customize factory:
> 
> ```python
> fctr = make_factory(Srvc)
> 
> def factory(svcs_container):
>     with fctr(svcs_container) as instance:
>         yield instance
> 
> registry.register(Srvc, factory)
> ```
> 
> In case our service has it's own life circle an supposed to be used as context manager.

I don't quite follow here. It kinda sounds like feature-creep? I'm trying to take advantage of type annotations here and not automate away everything to the point of svcs becoming an opaque blog like most DI frameworks. 😅

**Comment by proofit404** [comment]
![image](https://github.com/user-attachments/assets/03c3431a-d708-4dfb-871b-fe3a14e674dd)

@hynek let me put it this way :palm_tree: :cocktail:

**Comment by proofit404** [comment]
@hynek I would like to contribute this feature to the project.

Any thoughts on constraints / design decisions mentioned above?

**Reply by hynek** [reply]
Cool!! I've been meaning to attack the bigger issues in *svcs* forever but I keep getting pulled into my Big Two (attrs, structlog) all the time. :(

**Reply by proofit404** [reply]
I'll be hiking in the mountains for next two weeks. Planning to make some progress on this feature before new year.

**Reply by hynek** [reply]
looong hike @proofit404! 😇

**Reply by proofit404** [reply]
I had an overlap with team-building event, and family visit for a new year. Actually, today is my first working day since November.

**Reply by pauleveritt** [reply]
Funny enough, I loaded this stuff (and my thing) back into my brain yesterday.

**Comment by pauleveritt** [comment]
@proofit404 If you do wind up working on it, I have code/tests for sniffing that I could look at in helping.

**Comment by proofit404** [comment]
@hynek Hi, finally get time to work on this feature.

What would be final api design decision?

```python
registry.register_factory(Wrapper, svcs.AutoSvc(Wrapper))
```

**Reply by hynek** [reply]
Maybe we can even go as far as `Auto`? 🤔 Renaming it before the first release won’t be a big deal until we release it. If y’all have thoughts I’m listening.

**Reply by proofit404** [reply]
I would prefer lowercase `auto`, if possible.

**Reply by hynek** [reply]
Call it auto and the class that presumably gets instantiated AutoService then?

**Reply by proofit404** [reply]
Why not. If I understand correctly, `AutoService` would be an internal thing?

**Reply by hynek** [reply]
Yeah mostly. We’ve got precedents with ServicePing and RegisteredService that allow for limited introspection. Make it definitely frozen, tho.

**Reply by hynek** [reply]
@proofit404 no rush, but I've noticed that I've been carrying around a bunch of bug fixes that never got released. do you think it makes sense to wait for you? there'll definitely be more releases this year.

**Reply by proofit404** [reply]
I definitely don't want to be blocker here. Please, release fixes you have.

**Comment by proofit404** [comment]
**Status update:** finally got a chance to encorporate svcs into project at work. Hope it will land well, so I wold have a code base to test feature.

**Reply by pauleveritt** [reply]
@proofit404 If you're still working on the injector code for signature sniffing...I have some code. It's only mediocre, but maybe the test cases are helpful.

**Comment by pauleveritt** [comment]
@hynek Claude (with some heavy rules and steering) and I have [an implementation](https://github.com/pauleveritt/svcs-di/blob/main/src/svcs_di/auto.py). Some notes:

- This project targets 3.14 but it's straightforward to target 3.12
- I kept that file as free of scope-creep as I could
- It has a companion `auto.pyi` for reasons that are explained in the package
- It has one architecture astronauting
  - It gets the injector from the registry
  - This allows alternate injectors
  - I have two in this package
  - Spot your name in the README and you'll find more details in the note I left you

**Reply by hynek** [reply]
It's available on PyPI now too, so everyone interested in this topic should check it out and provide feedback and maybe we can fold it into *svcs* proper eventually (pending _attrs_ support, of course 🤓): https://pauleveritt.github.io/svcs-di/

**Reply by pauleveritt** [reply]
That's a good point about attrs support. I'm using it more heavily in my related, more-ambitious thing. But this package definitely should support it. I'll work on a revision.

I think I also need to do some content on svcs itself. I'm starting to see that svcs and SOA (and Harry's book) has a role in how agentic engineering might land.

**Comment by g0di** [comment]
I'm using svcs a lot recently and I found a pretty easy implementation of "auto wiring" based on type annotations. This avoid us to write many factories for building our various services. Maybe this could be a "simple" utility function that could be added to the library. Here is a working example:

```python
from collections.abc import Callable
from inspect import isclass
from typing import TypeVar, get_type_hints

from svcs import Container, Registry


T = TypeVar("T")


def autowire(fn_or_cls: Callable[..., T]) -> Callable[[Container], T]:
    def wrapper(svcs_container: Container) -> T:
        constructor = fn_or_cls.__init__ if isclass(fn_or_cls) else fn_or_cls
        # include_extra=True to avoid get_type_hints from unwrapping Annotated
        type_hints = get_type_hints(constructor, include_extras=True)
        type_hints.pop("return", None)

        return fn_or_cls(
            **{
                name: container.get(type_)
                for name, type_ in type_hints.items()
            }
        )

    return wrapper


class Foo:
    def __init__(self, bar: "Bar") -> None:
        self.bar = bar

    def __repr__(self) -> str:
        return f"Foo(bar={self.bar!r})"


class Bar:
    def __init__(self, value: str) -> None:
        self.value = value

    def __repr__(self) -> str:
        return f"Bar(value={self.value!r})"


@autowire
def create_bar(value: str) -> Bar:
    return Bar(value)


registry = Registry()
registry.register_value(str, "Hello, World!")
registry.register_factory(Foo, autowire(Foo))
registry.register_factory(Bar, create_bar)


with Container(registry) as container:
    print(repr(container.get(Foo)))
```

As you can see, it handles both classes and functions. The "autowire" function is rather simple and does the job. It could be further enhanced to handle positional only arguments and/or default values without too much trouble:
- Using `inspect.signature` one can detect if an argument is positional only so our autowire function can pass it this way (this is something I already did)
- Using `inspect.signature` one can get default values of parameters. We could try to get the parameter value from the container if it is registered or leave the default value if not

 As this is an utility function, it does not have any impact on the core of svcs and can be opt-in/opt-out easily.

@hynek If this is something that could fit the library I can propose a PR.

**Reply by hynek** [reply]
this looks a bit closer to what I had originally in mind and I was bit surprised by the size/complexity of @pauleveritt 's solution. I'm super swamped ATM, so could y'all explain the trade-offs between the two? It's a bit tough to grasp given it's mostly meta-programming :D

(pls no LLM output 🥺)

**Reply by g0di** [reply]
Paul's solution does the same as mine using a same public API (a single `auto()` decorator) plus it manages async and default values of factories. The main difference (except code complexity that I'll explain later) is that Paul's expect types to be annotated with an `Inject[Bar]` Annotated type to tell an injector which types should be injected at resolution time where I try to resolve all parameters.

Here is my personal feedback:

- Paul's solution is heavier but he already mentionned it so in the end, its almost the same:
  > Note to Hynek: The following is custom to me, not for svcs directly, and it is all insane so stop reading here. Don’t look, your eyes will burn with all the energy from the Great Pyramid.
  So I think his idea was more to try out something rather than having this implemented in svcs.
- Paul's solution provides more features than me (async, default values handling, custom injector for further extensibility) but the main ones (async, default) could be managed by a simpler implementation as well
- I'm personally not a big fan of using a dedicated type to tell svcs which fields should be injected for two reasons:
  - You have to annotate your code and services with svcs custom types, which IMHO goes against svcs own description
    > It's suitable for implementing [Inversion of Control](https://svcs.hynek.me/en/latest/glossary.html#term-Inversion-of-Control) using either dependency injection or service location while not requiring global state, decorators, or mangling of function signatures.
  - I personally don't see the benefit of marking explicitly which fields has to be injected. Also, I did not get the behavior of resolving a callable which one his field has no default and is not annotated with `Inject`
  
 @pauleveritt don't hesitate to correct me if I'm wrong. No offense by the way, I'm not trying to critisize your work.

**Comment by pauleveritt** [comment]
@g0di  Thank you for a very considerate and thoughtful evaluation. I think you have it right. 

1. *`Inject[]`*. This is my third implementation of these ideas (wired, then Hopscotch, now this.) In the two previous I did *not* do a customer marker like `Inject[]`. I just implicitly did the container lookup. It felt later like other libraries in Python went away from that as being too magical. 

People have a distrust of DI. I wanted to make it clear: you don't get DI unless you ask for it. I would be fine with getting out of that for the thing that goes in `svcs`. 

2. *Custom injector*. I think my little universe can proceed on its own path if you removed all of that.

3. *Default values*. I think you should consider this, as it's kind of the Pythonic way.

I'm up for working with @g0di on his solution and retiring mine.

**Reply by hynek** [reply]
The thing with `Inject[]` is that, yes it's more explicit, but it also couples the dependency to the injector which I _think_ is generally not desirable? For instance, you'd have to bleed svcs into your repository if you wanted them to be instantiated with database connections.

I mean I personally don't think I would use either, because I just don't think it's too much work to type out how objects are constructed explicitly and to my taste the ratio of "keypresses saved" vs "reasoning complexity introduced / debugging made more difficult" is rather bad, so I might not be the best judge here.

**Reply by g0di** [reply]
I think that @pauleveritt is pretty fine with not using the `Inject[]` annotation in the end. I can work on a PR when I'll have some spare time if you're fine with my original approach.

**Comment by g0di** [comment]
I opened a PR as discussed so far: #167.

**Comment by hynek** [comment]
fixed in #167 and released in 26.1.0 – thanks everyone 🥹

**Reply by g0di** [reply]
Thank you for the hard work to finish and merge the PR!
